### ISSUE TITLE
Definition of eventDate returned by the API

### ISSUE THREAD

**Issue by cgendreau** [gbif/gbif-api#3]
### Definition of eventDate returned by the API
According to DarwinCore the [eventDate](http://rs.tdwg.org/dwc/terms/index.htm#eventDate) is, "the date-time or interval during which an Event occurred. For occurrences, this is the date-time when the event was recorded.".

This brings 2 concepts: 
 1) The date and time relative to the location of the event 
 2) The exact moment the event occurred on the timeline

To illustrate the example, lets take the worst case scenario:

eventDate : 2016-09-15T00:05:00+1400 (LINT, Kiritimati, Kiribati - Christmas Island, UTC+14)

At the exact same moment, it is:
 * 2016-09-14T10:05:00+0000 (UTC)
 * 2016-09-14T11:05:00+0100 (BST, London, United Kingdom, UTC+1)
 * 2016-09-13T00:05:00-1200 (AoE, Baker Island, US Minor Outlying Islands, UTC-12)
 
In the worst case, a single eventDate 2016-09-15 can actually be on 3 different dates depending where you are on Earth.

This raises some questions:
 * When a user search for all records that occurred on 2016-09-14, what should be returned?
 * When no timezone is provided, which one should be used?
 

The tradeoff solution that answer the 2 questions above is to handle eventDate as a local date time. This means the the eventDate is relative to its location and the time zone is therefore ignored. When you query GBIF for a specific date, you then get all records that were recorded on that date relatively to their location. The counter side of such solution is that it makes it very difficult to ordered eventDate based on the timeline (in the order they exactly happened in the real world) with a precision higher than 26 hours (biggest difference between 2 timezones). 

Please let us know if you have good reason to think that handling eventDate as local date time could affect you program/work.

**Comment by mdoering** [comment]
I very much agree that eventDate should be treated as the local datetime relative to the timezone where it was recorded. And searching occurrences should also be done using local datetimes, so a query for 31.January 1973 at noon would match records that actually happened after each other in GMT time.

Should we add the correct timezone during eventDate interpretation depending on the location we determined?

**Comment by MattBlissett** [comment]
It seems most likely that a collection database would record the local date: with this example, the British botanist on a trip to Kiribati writes "15 September" in their field book, and when they return to London they'll put "15 September" into their database. Where we only have a date (no time) this is probably appropriate.

If we have a time it will be more complicated.  It could be significant that something was observed at night, or at dawn.  What kind of queries do users need?

Civil timezones aren't useful between countries -- Tajikistan and China share a border, but the time zone difference is five hours. Does anyone fancy implementing local solar time for searches 😁 ? (Is it a straightforward calculation from the position?)

Civil timezones also change. Kiribati was at -11 and -10 until 1995, when they switched to +13 and +14.  The Java timezone library will easily handle this, but we need to know what we're trying to achieve.

**Comment by tucotuco** [comment]
I vote for searching on the date and time given (not necessarily local time - some records might be recorded as UTC regardless of where they are, many have no time zone indicator and are therefore ambiguous), not the time on a timeline. Users who need that specialized situation can do so (or not, depending on what they get) from searches broad enough to get everything they might be interested in and create their own timeline.

**Comment by cgendreau** [comment]
I agree that the "local date time" I mentioned above is actually the date time provided. Thanks @tucotuco for the clarification.

**Comment by niconoe** [comment]
I think handling the eventDate has a local time (except if stated otherwise, i.e. UTC) is:

- probably enough for 99%+ of uses
- has the advantage of staying rather simple
- and to correctly interpret dates what the scientist meant in most cases, like @MattBlissett mentioned.

**Comment by peterdesmet** [comment]
Not sure I fully understand, but are the different ideas:

* Handle all dates as provided, so searching on 2012-03-04T15:00 will get you all occurrences locally collected at 3PM, so you know you're getting records collected in the afternoon.
* Translate all dates to UTC dates (using provided timezone info or location). Searching on a 2012-03-04T15:00 implies using UTC, so you can get morning, afternoon, or night occurrences, depending on where they were collected.

Is that a correct? If so, the first search is more intuitive, but with 2 remarks:

* How are dates with timezone info (e.g. dates are in UTC, like the gull tracking data) handled? Are these transformed to a local time (i.e. in the other direction)?
* Maybe it's still an option to translate all dates to UTC (cf. coordinates), provide this info in downloads and use issues like UTC_DATE_CALCULATED_FROM_LOCATION. It should then be possible to query using the intuitive local time + provide info for users who want all occurrence on a UTC timeline.

Note: one usecase where precise times are important are tracking data. Which is why we provide them in UTC. As Nico says, for most other use cases it does not matter.

**Comment by cgendreau** [comment]
@peterdesmet local date time is not the best word but I don't have a better one for now. The idea behind is to simply ignore the timezone, no transformation would be applied, so it is closer to the first point you mentioned.

**Comment by peterdesmet** [comment]
Ok, but if no dates will be transformed, there is still the slight issue that if you assume you're querying for 3PM (local time) all around the world, that dates provided in UTC won't match that expectation. Correct?

**Comment by cgendreau** [comment]
yes, this is correct. I have to agree that the sentence "This means the eventDate is relative to its location and the time zone is therefore ignored." is misleading since it assumes the eventDate provided always refers to the local date time and has not been translated to UTC.

For the moment, if a date is provided in UTC it will be used as is. We do not plan to translate them to the "local date time" of the event. That's why "local date time" is not the right word. But I understand there might be a need for it.

Currently it is not possible to search by "time", but the definition would be more "query all records that provided 3PM as time".

**Comment by MattBlissett** [comment]
Does it make sense to query for 3PM local time, all around the world?  That means afternoon in Beijing, but still morning in Western China.  It means things an hour apart from each other across the border between US states, or even counties.  3AM sometimes means a time that doesn't exist (when the country jumps forward an hour for summer time) or exists twice (when they go back).  Even travelling due South, it means times an hour or more apart for places like England and France, or China and India.  Many places are in the "wrong" timezone: 

![Map](https://upload.wikimedia.org/wikipedia/commons/4/4b/Solar_time_vs_standard_time.png)

(White is where civil and solar time coincide, dark red/blue is where they are especially "wrong".)

I think anyone whose research needs this much detail regarding times will need to query for a wider range of data, and do some processing themselves.  Users could convert a UTC time into a local solar time based on the longitude, and then they'd be able to get all observations in the afternoon, or at sunrise.

So long as we preserve what's provided to us, and don't make it less accurate than it was before, it should be possible to do this analysis.

I'm reminded of http://infiniteundo.com/post/25509354022/more-falsehoods-programmers-believe-about-time

**Comment by mdoering** [comment]
Looking at the map I wonder how times are handled on boats in the oceans. Do people regularly adjust clocks as the ship moves?

**Comment by MattBlissett** [comment]
@mdoering: Yes, ships usually change their clock every day or so. This is why even the uninhabited time zones exist in databases.

A search of our verbatim data shows very few people provide a timezone — iNaturalist and two datasets from New Zealand account for the overwhelming majority.

However, the NZ ones are both machine sample datasets, so it does seem wrong to mess them up. At the moment we do this: `2008-01-15T07:11+12` → `2008-01-15T07:11:00.000+0000`.

I agree with Peter that would be better as `2008-01-15T07:11+12` → `2008-01-15T07:11:00` — it is no longer wrong. Data remains comparable within its own dataset, but isn't comparable with a nearby observation from the same minute recorded as `2008-01-14T19:11+00`.  However, its not a trivial change to the API.

One of those NZ datasets: https://www.gbif.org/dataset/2a636fd6-dc73-49c0-9e5b-26bd05532313
